### PR TITLE
Implement app creation in portal

### LINKED ISSUE #354: Improve portal support of multi-tenancy
- Front-end
  - Improve UI of app selection page
    - Show public origin in app list
    - Create app button + dialog
- Back-end
  - Create app API
    - Create ConfigMap & Secret in k8s
    - Generate secret YAML using same logic as CLI
      - Refactor CLI config generation logic to core config package
    - Accept cluster-wide credentials in environment variable
      - Database URL & schema, Redis URL, SMTP, etc.
  - Secret config query
    - redact secret item if exact match with cluster-wide credentials
  - Secret config mutation
    - use cluster-wide credentials for secret item if missing/same as redacted value

### PR DESCRIPTION
**Pull request by kiootic**
fix #354

### CODE REVIEW

**Review by louischan-oursky — COMMENTED** [review]
We should update .env.example so that `cp .env.example .env` should work out of the box.

This is what I used now

```
APP_HOST_TEMPLATE={{ .AppID }}.localhost:3000
APP_SECRET_DATABASE_URL=postgres://postgres:postgres@127.0.0.1:5432/postgres?sslmode=disable
APP_SECRET_DATABASE_SCHEMA=public
APP_SECRET_REDIS_URL=redis://localhost
```

File: pkg/portal/service/config.go [hunk 1]
```diff
@@ -151,3 +169,14 @@ func (s *ConfigService) updateLocalFS(l *configsource.LocalFS, appID string, upd
 
 	return nil
 }
+
+func (s *ConfigService) createKubernetes(k *configsource.Kubernetes, appID string, appConfigYAML []byte, secretConfigYAML []byte) error {
+	_, err := k.ResolveContext(appID)
+	if err != nil && !errors.Is(err, configsource.ErrAppNotFound) {
+		return err
+	} else if err == nil {
+		return ErrDuplicatedAppID
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/portal/service/config.go:180]: > But we need to create more than 1 resources and the API is not transactional :(

Yes, we'll need to handle cleaning up orphaned resources when creation failed unexpectedly as stated in TODO.

> Store everything as a secret. So the secret effectively is a FS for a particular app.

That sounds fine to me, let me open an issue for that.

> Do not use a single configmap for host resolution. Instead, we use real Ingress resource.

So the portal backend would create the associated ingress resource? Then developer would need to configure a template for the ingress.

File: pkg/portal/config/app.go [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	SMTPHost         string `envconfig:"SMTP_HOST"`
+	SMTPPort         int    `envconfig:"SMTP_PORT"`
+	SMTPMode         string `envconfig:"SMTP_MODE"`
+	SMTPUsername     string `envconfig:"SMTP_USERNAME"`
+	SMTPPassword     string `envconfig:"SMTP_PASSWORD"`
+	TwilioAccountSID string `envconfig:"TWILIO_ACCOUNT_SID"`
+	TwilioAuthToken  string `envconfig:"TWILIO_AUTH_TOKEN"`
+	NexmoAPIKey      string `envconfig:"NEXMO_API_KEY"`
+	NexmoAPISecret   string `envconfig:"NEXMO_API_SECRET"`
+}
```
[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/portal/config/app.go:27]: I suppose for purpose of branding, cluster admin would also want to have default custom email template/messages and web UI theme?

Currently cluster admin should be able to mount a customized default template directory into authgear docker image to achieve this. Let me open an issue for populating default per-app templates.

File: pkg/portal/service/config.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/portal/service/config.go:180]: > So the portal backend would create the associated ingress resource?

Yes

> Then developer would need to configure a template for the ingress.

More details?

[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/portal/service/config.go:180]: For example, some cluster may use nginx ingress controller, some would like to use other. Some may have single wildcard cert, some may use cert-manager + let's encrypt. All these differences need to customize the ingress resource for each app domain.

[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/portal/service/config.go:180]: But the template should be global. So it is not provided by the developer. It is provided by us.

[the remark this answers is not in the document]
Comment (kiootic) [reply pkg/portal/service/config.go:180]: However the created ingress may be different depending on the setup of deployed cluster? Therefore the cluster admin need to customize the created ingress using something like YAML template?

[the remark this answers is not in the document]
Comment (louischan-oursky) [reply pkg/portal/service/config.go:180]: Yes.

The developer - the one who uses Authgear
Us, cluster admin - the one who deploys Authgear

**Comment by louischan-oursky** [comment]
Anyway we have opened the followup issues so I am going to merge this.

**Pull request merged**